domain/credential/state/credential_test.go

```diff
@@ -0,0 +1,408 @@
+	c.Assert(out, gc.IsNil)
+}
+
+func (s *credentialSuite) createCloudCredential(c *gc.C, st *State, credentialName, cloudName, userName string) cloud.Credential {
+	authType := cloud.AccessKeyAuthType
+	attributes := map[string]string{
+		"foo": "foo val",
+		"bar": "bar val",
+	}
+
+	s.addCloud(c, st, cloud.Cloud{
+		Name:      cloudName,
+		Type:      "ec2",
+		AuthTypes: cloud.AuthTypes{cloud.AccessKeyAuthType, cloud.UserPassAuthType},
+	})
+
+	cred := cloud.NewNamedCredential(credentialName, authType, attributes, false)
+	err := st.UpsertCloudCredential(ctx.Background(), credentialName, cloudName, userName, cred)
+	c.Assert(err, jc.ErrorIsNil)
+	return cred
+}
+
+func (s *credentialSuite) TestAllCloudCredentials(c *gc.C) {
+	st := NewState(testing.TxnRunnerFactory(s.TxnRunner()))
+
+	one := s.createCloudCredential(c, st, "foobar", "cirrus", "bob")
+	two := s.createCloudCredential(c, st, "foobar", "stratus", "bob")
+
+	// Added to make sure it is not returned.
+	s.createCloudCredential(c, st, "foobar", "cumulus", "mary")
+
+	out, err := st.AllCloudCredentials(ctx.Background(), "bob")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(out, jc.DeepEquals, []CloudCredential{
+		{Credential: one, CloudName: "cirrus"},
+		{Credential: two, CloudName: "stratus"},
+	})
+}
+
+func (s *credentialSuite) TestInvalidateCloudCredential(c *gc.C) {
+	st := NewState(testing.TxnRunnerFactory(s.TxnRunner()))
+
+	one := s.createCloudCredential(c, st, "foobar", "cirrus", "bob")
+	c.Assert(one.Invalid, jc.IsFalse)
+
+	ctx := ctx.Background()
+	reason := "testing, testing 1,2,3"
+	err := st.InvalidateCloudCredential(ctx, "foobar", "cirrus", "bob", reason)
+	c.Assert(err, jc.ErrorIsNil)
+
+	updated, err := st.CloudCredential(ctx, "foobar", "cirrus", "bob")
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(updated.Invalid, jc.IsTrue)
+	c.Assert(updated.InvalidReason, gc.Equals, reason)
+}
+
+func (s *credentialSuite) TestInvalidateCloudCredentialNotFound(c *gc.C) {
+	st := NewState(testing.TxnRunnerFactory(s.TxnRunner()))
+
+	ctx := ctx.Background()
+	err := st.InvalidateCloudCredential(ctx, "foobar", "cirrus", "bob", "reason")
+	c.Assert(err, jc.Satisfies, errors.IsNotFound)
+}
+
+type watcherFunc func(namespace, changeValue string, changeMask changestream.ChangeType) (watcher.NotifyWatcher, error)
+
+func (f watcherFunc) NewValueWatcher(
+	namespace, changeValue string, changeMask changestream.ChangeType,
+) (watcher.NotifyWatcher, error) {
+	return f(namespace, changeValue, changeMask)
+}
+
+func (s *credentialSuite) watcherFunc(c *gc.C, expectedChangeValue string) watcherFunc {
+	return func(namespace, changeValue string, changeMask changestream.ChangeType) (watcher.NotifyWatcher, error) {
+		c.Assert(namespace, gc.Equals, "cloud_credential")
+		c.Assert(changeMask, gc.Equals, changestream.All)
+		c.Assert(changeValue, gc.Equals, expectedChangeValue)
+
+		factory := testing.StubFactory(s.TxnRunner(), s.events)
```

Comment:
Actually, I rescind my prior comment about the factory. Just delete it. This is the only usage, and you invoke it directly, checking an error that is never nil.

Given it's specificity, I'd rather put the implementation of `stubWatchableDB` here instead of the testing package, and just do
```go
db := &stubWatchableDB{
    TxnRunner: runner,
    events:    events,
}
base := eventsource.NewBaseWatcher(db, ...
```